lualib/resty/events/worker.lua

```diff
@@ -143,69 +170,63 @@ function _M:communicate(premature)
         return
     end
 
-    self._connected = true
     log(DEBUG, _worker_id, " on (", listening, ") is ready")
 
+    self._connected = true
+
     local read_thread = spawn(function()
-        while not exiting() do
+        while not terminate(self) do
             local data, err = conn:recv_frame()
-
-            if exiting() then
-                return
-            end
-
             if err then
                 if not is_timeout(err) then
-                    return nil, err
+                    return err_if_not_exiting("failed to read event", err)
                 end
 
                 -- timeout
                 goto continue
             end
 
             if not data then
-                return nil, "did not receive event from broker"
+                log_if_not_exiting(ERR, "did not receive event from broker")
+                goto continue
             end
 
-            local d, err = decode(data)
-            if not d then
-                return nil, "failed to decode event data: " .. err
+            local event_data, err = decode(data)
+            if err then
+                log_if_not_exiting(ERR, "failed to decode event data: ", err)
+                goto continue
             end
 
             -- got an event data, push to queue, callback in events_thread
-            local ok, err = self._sub_queue:push(d)
-            if not ok then
-                log(ERR, "failed to store event: ", err, ". ",
-                         "data is :", cjson_encode(d))
+            local _, err = self._sub_queue:push(event_data)
+            if err then
+                log_if_not_exiting(ERR, "failed to store event: ", err, ". ",
+                                        "data is: ", cjson_encode(event_data))
             end
 
             ::continue::
-        end -- while not exiting
-    end)  -- read_thread
+        end -- while not terminate
+
+        return true
+    end) -- read_thread
 
     local write_thread = spawn(function()
         local counter = 0
 
-        while not exiting() do
+        while not terminate(self) do
             local payload, err = self._pub_queue:pop()
-
-            if not payload then
+            if err then
                 if not is_timeout(err) then
-                    return nil, "semaphore wait error: " .. err
+                    return err_if_not_exiting("semaphore wait error", err)
                 end
 
                 -- timeout
                 goto continue
             end
 
-            if exiting() then
-                return
-            end
```

Comment:
Should we remoe this? The `pop()` above may take time.

Reply:
It may take a time (1 second I think), but I wanted to have consistent behavior. If event is long running, it should yield, like `reconfigure` does. I want here to leave the process in consistent state, e.g. no events are skipped and I left that logic to:

```
    if exiting() then
        kill(read_thread)
        kill(write_thread)
        kill(events_thread)
        return
    end
```

Perhaps later on we want to change that to drain events when exiting. And if that is needed it is easier to do it there at the end.